### ISSUE TITLE
Improving the macOS Build Instructions

### ISSUE THREAD

**Issue by elsiehupp** [nextcloud/desktop#2893]
> I'd be glad to have a PR to improve the documentation though.

_Originally posted by @er-vin in https://github.com/nextcloud/desktop/issues/2888#issuecomment-770812661_

I tried editing the wiki. It isn't editable by non-collaborators, it isn't part of the repository itself, and it can't be forked like a normal github repository, so I can't edit it on my own and then do a pull request. I tried following [these instructions](https://gist.github.com/omaraboumrad/35654da0a376c57a2e0ab4d92ad0c339) and got this error:
```
% git remote add upstream https://github.com/nextcloud/desktop.wiki.git
fatal: not a git repository (or any of the parent directories): .git
```
So for now improving the wiki seems to be a non-starter. I'll put my draft in this Issue, though.

EDIT: Note: this is a follow-up to #2856, but I felt like it would be preferable to treat this as separate since it's a quasi-pull-request for the Wiki.

**Comment by elsiehupp** [comment]
This is a bit long, so FYI it's broken into three sections:
1. Existing build instructions with annotations
2. Build warnings with annotations
3. An updated version of the build instructions, and
4. Potential changes to CMakeLists.txt

## Let's try the official build instructions! ##
> Note: [The macOS build instructions from the Nextcloud Desktop GitHub wiki](https://github.com/nextcloud/desktop/pull/2888/) are quite sad, so I'm using [the instructions from the Nextcloud Client Manual](https://docs.nextcloud.com/desktop/3.1/building.html) instead.

1. Install Xcode:
 - https://developer.apple.com/xcode/
Or:
 - https://apps.apple.com/app/xcode/id497799835
2. Install Xcode command line tools::
```
% xcode-select –install
```
3. Install Homebrew from https://brew.sh (which will just give you the following)
```
% /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```
4. Add the Nextcloud repository:
 Assuming you've forked the repository:
 `brew tap <username>/desktop git@github.com:<username>/desktop.git`
 Or HTTPS:
 `% brew tap <username>/desktop`
 Or the upstream repository (read-only if you're not a collaborator):
 `% brew tap nextcloud/desktop git@github.com:nextcloud/desktop.git`
 Or `brew tap nextcloud/desktop`
> You might find `/usr/local/Homebrew/Library/Taps/<username>/homebrew-desktop` a slightly inconvenient place for the cloned repository, in which case:
> `brew install git` (if you haven't already)
> Then, in the more convenient folder:
> `git clone git@github.com:<github_username>/desktop.git`
> Or:
> `git clone https://github.com/<github_username>/desktop.git`
> Or, read-only:
> `git clone git@github.com:nextcloud/desktop.git`
> Or:
> `git clone https://github.com/nextcloud/desktop.git`

5. Install dependencies:
``` 
% brew install $(brew deps owncloud-client)
```

> Yeesh, maybe you guys could create your own Homebrew formula? Depending on OwnCloud's seems a bit iffy... This part is straight from the official documentation!
>
> Note that it only installs the following, when you could stand to have more:
> ```
> % brew deps owncloud-client
> qt
> qt5
> qtkeychain
> ```
> Also:
> ```
> qt is keg-only, which means it was not symlinked into /usr/local,
> because Qt 5 has CMake issues when linked.
>
> If you need to have qt first in your PATH, run:
>   echo 'export PATH="/usr/local/opt/qt/bin:$PATH"' >> ~/.zshrc
>
> For compilers to find qt you may need to set:
>   export LDFLAGS="-L/usr/local/opt/qt/lib"
>   export CPPFLAGS="-I/usr/local/opt/qt/include"
> ```
> Well, that's nice.

6. Installing Qt is unnecessary because of the above step, but otherwise it would be:
```
% brew install qt
```
> Note that you could install the full Qt developer package, but the dependencies are sufficient for now.

7. Adding Qt to the path is unnecessary because Homebrew did that for us, but it's the equivalent of:
```
% brew link qt
```

8. Install QtKeychain:
```
% brew install qtkeychain
```
> Note that this is necessary despite Step 5. I don't know why.

9. Create build directory:
```
% cd /usr/local/Homebrew/Library/Taps/<github_username>/homebrew-desktop
```
(Or your preferred directory)
```
% mkdir build
```
10. Install the Packages package creation tool:
 http://s.sudre.free.fr/Software/Packages/about.html
> I don't know if this is strictly necessary for a test build

11. Enable git submodules:
```
% cd desktop 
% git submodule init
% git submodule update
```
> I honestly don't know what this does.

12. Generate the build files:
```
% cd build 
% cmake .. -DCMAKE_INSTALL_PREFIX=~/nextcloud-desktop-client -DCMAKE_BUILD_TYPE=Debug -DNO_SHIBBOLETH=1
zsh: command not found: cmake
```
> Oh. Install CMake:
> ```
> % brew install cmake
> ```
> ...
> ```
> Error: An unexpected error occurred during the `brew link` step
> The formula built, but is not symlinked into /usr/local
> Permission denied @ apply2files - /usr/local/share/emacs/site-lisp/cmake
> Error: Permission denied @ apply2files - /usr/local/share/emacs/site-lisp/cmake
> ```
> This is [a known issue](https://stackoverflow.com/a/63241724). Do the following to fix:
> ```
> % sudo chown -R $(whoami):admin /usr/local/* \
> && sudo chmod -R g+rwx /usr/local/*
> ```
> Note that changing the permissions this way may induce other shell warnings (i.e. from `oh-my-zsh`).
>
> Now try again:
> ```
> % brew link cmake
> ```
> And:
> ```
> % cmake .. -DCMAKE_INSTALL_PREFIX=~/nextcloud-desktop-client -DCMAKE_BUILD_TYPE=Debug -DNO_SHIBBOLETH=1
> ```
> ...
> ```
> CMake Error at /usr/local/Cellar/cmake/3.19.4/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:218 (message):
>   Could NOT find OpenSSL, try to set the path to OpenSSL root folder in the
>   system variable OPENSSL_ROOT_DIR (missing: OPENSSL_CRYPTO_LIBRARY
>   OPENSSL_INCLUDE_DIR) (Required is at least version "1.1")
> ```
> macOS uses LibreSSL, which doesn't work for this purpose, so install OpenSSL:
> ```
> % brew install openssl
> ```
> ...
> ```
> openssl@1.1 is keg-only, which means it was not symlinked into /usr/local,
> because macOS provides LibreSSL.
>
> If you need to have openssl@1.1 first in your PATH, run:
>   echo 'export PATH="/usr/local/opt/openssl@1.1/bin:$PATH"' >> ~/.zshrc
>
> For compilers to find openssl@1.1 you may need to set:
>   export LDFLAGS="-L/usr/local/opt/openssl@1.1/lib"
>   export CPPFLAGS="-I/usr/local/opt/openssl@1.1/include"
> ```
> Huh. Well, CMake still gives me the same—because OpenSSL via Homebrew didn't get linked to the `PATH`, but doing the following clears it up:
> ```
> % export OPENSSL_ROOT_DIR=$(brew --prefix openssl)
> ```
> Note that this will need to be re-run, which is why I added it to CMakeLists.txt.
>
> Re-running CMake gets me:
> ```
> CMake Error at src/CMakeLists.txt:7 (find_package):
>   By not providing "FindQt5.cmake" in CMAKE_MODULE_PATH this project has
>   asked CMake to find a package configuration file provided by "Qt5", but
>   CMake did not find one.
>
>   Could not find a package configuration file provided by "Qt5" (requested
>   version 5.12) with any of the following names:
>
>     Qt5Config.cmake
>     qt5-config.cmake
>
>   Add the installation prefix of "Qt5" to CMAKE_PREFIX_PATH or set "Qt5_DIR"
>   to a directory containing one of the above files.  If "Qt5" provides a
>   separate development package or SDK, be sure it has been installed.
> ```
> Doing this clears that up:
> ```
> % export Qt5_DIR=$(brew --prefix qt)
> ```
> Note, again, that this will need to be re-run, which is why I added it to CMakeLists.txt
>
> Finally, the build files have been installed, though I still have plenty of build warnings.

11. Enable git submodules:
```
% cd desktop 
% git submodule init
% git submodule update
```
> I honestly don't know what this does.

12. Generate the build files:
```
% cd build 
% cmake .. -DCMAKE_INSTALL_PREFIX=~/nextcloud-desktop-client -DCMAKE_BUILD_TYPE=Debug -DNO_SHIBBOLETH=1
```
> There are still non-fatal build warnings, but at this point the build does succeed.

13. Compile and install:
```
% make install
```
> This shows tons and tons of compiler warnings, but the resulting application seems to work perfectly fine, so I will treat them as beyond the scope.

## Let's go through the build warnings! ##
```
CMake Warning (dev) at /usr/local/Cellar/cmake/3.19.4/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:426 (message):
  The package name passed to `find_package_handle_standard_args` (INOTIFY)
  does not match the name of the calling package (INotify).  This can lead to
  problems in calling code that expects `find_package` result variables
  (e.g., `_FOUND`) to follow a certain pattern.
```
> Presumably the character case in CMakeLists.txt is inconsistent.

```
-- Could NOT find INOTIFY
-- Could NOT find Sphinx
```
> These are inapplicable to macOS and indeed should be put within an `if(NOT APPLE)` block in CMakeLists.txt

```
-- Could NOT find PdfLatex
```
> PdfLatex is available for macOS as MacTex or BasicTex. BasicTex includes only the basic packages and can be downloaded from:
> - https://www.tug.org/mactex/morepackages.html
>
> Or, direct link:
> - http://mirror.ctan.org/systems/mac/mactex/BasicTeX.pkg

```
-- Could NOT find PkgConfig
-- Could NOT find GLib2
```
> These can be installed from Homebrew:
> ```
> % brew install pkgconfig glib

```
CMake Warning (dev) at /usr/local/Cellar/cmake/3.19.4/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:426 (message):
  The package name passed to `find_package_handle_standard_args`
  (LIBCLOUDPROVIDERS) does not match the name of the calling package
  (Libcloudproviders).  This can lead to problems in calling code that
  expects `find_package` result variables (e.g., `_FOUND`) to follow a
  certain pattern.
```
> Again, presumably the character case in CMakeLists.txt is inconsistent.

```
-- Could NOT find LIBCLOUDPROVIDERS
```
> This is also inapplicable to macOS and indeed should be put within an `if(NOT APPLE)` block in CMakeLists.txt

```
CMake Warning at src/gui/CMakeLists.txt:203 (find_package):
  By not providing "FindQt5LinguistTools.cmake" in CMAKE_MODULE_PATH this
  project has asked CMake to find a package configuration file provided by
  "Qt5LinguistTools", but CMake did not find one.

  Could not find a package configuration file provided by "Qt5LinguistTools"
  with any of the following names:

    Qt5LinguistToolsConfig.cmake
    qt5linguisttools-config.cmake

  Add the installation prefix of "Qt5LinguistTools" to CMAKE_PREFIX_PATH or
  set "Qt5LinguistTools_DIR" to a directory containing one of the above
  files.  If "Qt5LinguistTools" provides a separate development package or
  SDK, be sure it has been installed.
```
> Let's try:
> ```
> % export Qt5LinguistTools_DIR=$(brew --prefix qt)/lib/cmake/Qt5LinguistTools/
> ```

```
-- Could NOT find Doxygen
-- Could NOT find CMocka
```
> These can also be installed from Homebrew:
> ```
> % brew install doxygen cmocka
> ```

```
CMake Warning:
  Manually-specified variables were not used by the project:

    NO_SHIBBOLETH
```
> I suppose the `-DNO_SHIBBOLETH=1` wasn't necessary, then?

## Let's simplify the instructions! ##

1. Install Xcode:
 https://apps.apple.com/app/xcode/id497799835

2. Install Xcode command line tools::
```
% xcode-select –install
```
3. Install Homebrew from https://brew.sh (which will just give you the following)
```
% /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```
4. Fix Homebrew permissions. (This is [a known issue](https://stackoverflow.com/a/63241724), and, yes, the workaround may lead to other shell warnings.)
```
% sudo chown -R $(whoami):admin /usr/local/* \
&& sudo chmod -R g+rwx /usr/local/*
```
5. Install Homebrew packages:
```
% brew install git qt qt5 qtkeychain cmake openssl pkgconfig glib doxygen cmocka
```

6. Add unlinked install paths to shell profile:
```
% echo 'export OPENSSL_ROOT_DIR="/usr/local/opt/openssl@1.1"' >> ~/.zshrc
% echo 'export Qt5_DIR="/usr/local/opt/qt"' >> ~/.zshrc
% echo 'export Qt5LinguistTools_DIR="/usr/local/opt/qt/lib/cmake/Qt5LinguistTools/"' >> ~/.zshrc
```
> If the reader uses `bash` instead of `zsh`, they would use a different profile than `~/.zshrc`. As an aside, I'm fairly certain the `%` instead of a `$` signifies that the shell is `zsh` rather than `bash`.

> Note: I feel like it would be tidier and more resilient to update CMakeLists.txt so that it uses Homebrew to find these.

6. Install BasicTex (or MacTex)
 https://www.tug.org/mactex/morepackages.html
 Or, direct link:
 http://mirror.ctan.org/systems/mac/mactex/BasicTeX.pkg

7. Install the Packages package creation tool:
 http://s.sudre.free.fr/Software/Packages/about.html

10. Clone the Nextcloud repository to a convenient location, such as `~/Repositories`:
```
% cd ~/Repositories
```
Then, assuming you've already forked Nextcloud Desktop:
```
git clone git@github.com:<github_username>/desktop.git
```
Or, for HTTPS (soon to be deprecated):
```
git clone https://github.com/<github_username>/desktop.git
```
Or, if you haven't forked Nextcloud Desktop, clone as read-only:
```
git clone git@github.com:nextcloud/desktop.git
```
Or, for HTTPS (soon to be deprecated):
```
git clone https://github.com/nextcloud/desktop.git
```

11. Enable git submodules:
```
% cd ~/Repositories/desktop
% git submodule init
% git submodule update
```
> I honestly don't know what this does.

12. Create build directory:
```
% cd ~/Repositories/desktop
% mkdir build
```

13. Generate the build files:
```
% cd ~/Repositories/desktop/build
% cmake .. \
-DCMAKE_INSTALL_PREFIX=~/nextcloud-desktop-client \
-DCMAKE_BUILD_TYPE=Debug
```
> Note: I repeatedly used the absolute path after `cd` because it's somewhat likely a reader might go back and retry the steps out of order.

## Things to be fixed elsewhere

1. Correct the character cases for INOTIFY and LIBCLOUDPROVIDERS in CMakeLists.txt.

2. Create an `if(NOT APPLE)` block in CMakeLists.txt around INOTIFY, Sphinx, and LIBCLOUDPROVIDERS

3. Possibly add an `if(APPLE)` block to CMakeLists.txt that uses Homebrew to find `OPENSSL_ROOT_DIR`, `Qt5_DIR`, and `Qt5LinguistTools_DIR`, something like:
```
if (APPLE)
  execute_process(COMMAND brew --prefix openssl
    OUTPUT_VARIABLE OPENSSL_ROOT_DIR
    OUTPUT_STRIP_TRAILING_WHITESPACE)
  # message(STATUS "OPENSSL_ROOT_DIR detected as ${OPENSSL_ROOT_DIR}")

  execute_process(COMMAND brew --prefix qt
    OUTPUT_VARIABLE QT_PREFIX
    OUTPUT_STRIP_TRAILING_WHITESPACE)
  # message(STATUS "QT_PREFIX detected as ${QT_PREFIX}")

  set(Qt5_DIR "${QT_PREFIX}/lib/cmake/Qt5")
  # message(STATUS "Qt5_DIR detected as ${Qt5_DIR}")

  set(Qt5LinguistTools_DIR "${QT_PREFIX}/lib/cmake/Qt5LinguistTools/")
  # message(STATUS "Qt5LinguistTools_DIR detected as ${Qt5LinguistTools_DIR}")
endif()
```
> Note: I haven't actually tested the above code block. It's substantially pared down from what I put in my previous [pull request](https://github.com/nextcloud/desktop/pull/2888/).

**Comment by elsiehupp** [comment]
@er-vin @nicolasfella

**Comment by elsiehupp** [comment]
If the CMake script had a little more complexity, it would be better positioned to avoid conflicts. First, the entire thing could be wrapped in a conditional so that it only runs if it finds Homebrew:
```
find_program(HOMEBREW_EXECUTABLE brew)
if (HOMEBREW_EXECUTABLE)
...
endif()
```

Additionally, each variable could be wrapped in its own conditional so that CMake would only set the variable if the user had not already done so:
```
if(NOT OPENSSL_ROOT_DIR)
...
endif()
```
Altogether this becomes a bit elaborate (using `AND` operators to avoid further nesting):
```
if (APPLE)
  find_program(HOMEBREW_EXECUTABLE brew)
  if (HOMEBREW_EXECUTABLE AND NOT OPENSSL_ROOT_DIR)
    execute_process(COMMAND brew --prefix openssl
      OUTPUT_VARIABLE OPENSSL_ROOT_DIR
      OUTPUT_STRIP_TRAILING_WHITESPACE)
    # message(STATUS "OPENSSL_ROOT_DIR detected as ${OPENSSL_ROOT_DIR}")
  endif()

  if (HOMEBREW_EXECUTABLE AND NOT QT_PREFIX)
    execute_process(COMMAND brew --prefix qt
      OUTPUT_VARIABLE QT_PREFIX
      OUTPUT_STRIP_TRAILING_WHITESPACE)
    # message(STATUS "QT_PREFIX detected as ${QT_PREFIX}")
  endif()

  if (QT_PREFIX AND NOT Qt5_DIR)
    set(Qt5_DIR "${QT_PREFIX}/lib/cmake/Qt5")
    # message(STATUS "Qt5_DIR detected as ${Qt5_DIR}")
  endif()

  if (QT_PREFIX AND NOT Qt5LinguistTools_DIR)
    set(Qt5LinguistTools_DIR "${QT_PREFIX}/lib/cmake/Qt5LinguistTools/")
    # message(STATUS "Qt5LinguistTools_DIR detected as ${Qt5LinguistTools_DIR}")
  endif()
endif()
```
This could be further flattened:
```
if (APPLE)
  find_program(HOMEBREW_EXECUTABLE brew)
endif()

if (HOMEBREW_EXECUTABLE AND NOT OPENSSL_ROOT_DIR)
  execute_process(COMMAND brew --prefix openssl
    OUTPUT_VARIABLE OPENSSL_ROOT_DIR
    OUTPUT_STRIP_TRAILING_WHITESPACE)
  # message(STATUS "OPENSSL_ROOT_DIR detected as ${OPENSSL_ROOT_DIR}")
endif()

if (HOMEBREW_EXECUTABLE AND NOT QT_PREFIX)
  execute_process(COMMAND brew --prefix qt
    OUTPUT_VARIABLE QT_PREFIX
    OUTPUT_STRIP_TRAILING_WHITESPACE)
  # message(STATUS "QT_PREFIX detected as ${QT_PREFIX}")
endif()

if (HOMEBREW_EXECUTABLE AND QT_PREFIX AND NOT Qt5_DIR)
  set(Qt5_DIR "${QT_PREFIX}/lib/cmake/Qt5")
  # message(STATUS "Qt5_DIR detected as ${Qt5_DIR}")
endif()

if (HOMEBREW_EXECUTABLE AND QT_PREFIX AND NOT Qt5LinguistTools_DIR)
  set(Qt5LinguistTools_DIR "${QT_PREFIX}/lib/cmake/Qt5LinguistTools/")
  # message(STATUS "Qt5LinguistTools_DIR detected as ${Qt5LinguistTools_DIR}")
endif()
```
As before, I haven't yet tested this code anywhere, and I'm not sure where it would be inserted. I mean, plausibly, it could go in its own CMake module, but I don't know how to properly include it.

(For that matter, all of the `brew install` commands could go in a CMake module, but installing things outside of the build folder seems a bit sketch for the `cmake ..` command.)

My goal here isn't necessarily to reduce the complexity of setting up the build environment so much as avoiding forcing the user to repeatedly enter predictable CMake commands every time they build the project. Ideally, once everything is set up, the user should be able to get at least a working test build with just `cmake ..`, or, perhaps even more ridiculously, `cmake` (no `..`) in the project root directory.

**Comment by elsiehupp** [comment]
To further clarify: the main reason it's brittle to put the variables in the shell environment is that they are absolute paths, and if they are computed paths (i.e. using `brew --prefix [package]`), the computation has to run every time the shell restarts, i.e. when the user opens a new terminal window or tab. Manually entering the path-computation before every build is tedious, so creating a script computes the variables is more convenient. And telling the user to create a shell script or an environment file as part of the build instructions seems kind of silly, when actions like those are part of the main purpose of CMake.

The variables that CMake does not auto-detect in its current form seem to be related to limitations in Homebrew. I'm not entirely sure what the deal is with Homebrew Qt, though it does have a descriptive warning when you install it. OpenSSL, on the other hand, is by default not linked to the PATH because macOS ships with a conflicting implementation. Because the linking issues with Homebrew Qt and Homebrew OpenSSL seem to be structural and predictable, they would be good candidates for platform-specific CMake logic.

Other optional variables like `-DCMAKE_INSTALL_PREFIX` also seem to have default values that are geared towards Linux, so it seems reasonable that they could have `if(APPLE)` logic, as well.

**Comment by er-vin** [comment]
Thanks for all the work, interesting stuff, this is getting quite long though so for readability I will only quote reply on part three since this is what will lead to the wiki page we're contemplating updating.

> ## Let's simplify the instructions!
> 
>     1. Install Xcode:
>        https://apps.apple.com/app/xcode/id497799835
> 
>     2. Install Xcode command line tools::
> 
> 
> ```
> % xcode-select –install
> ```
> 
>     1. Install Homebrew from https://brew.sh (which will just give you the following)
> 
> 
> ```
> % /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
> ```
> 
>     1. Fix Homebrew permissions. (This is [a known issue](https://stackoverflow.com/a/63241724), and, yes, the workaround may lead to other shell warnings.)
> 
> 
> ```
> % sudo chown -R $(whoami):admin /usr/local/* \
> && sudo chmod -R g+rwx /usr/local/*
> ```
> 
>     1. Install Homebrew packages:
> 
> 
> ```
> % brew install git qt qt5 qtkeychain cmake openssl pkgconfig glib doxygen cmocka
> ```
> 
>     1. Add unlinked install paths to shell profile:
> 
> 
> ```
> % echo 'export OPENSSL_ROOT_DIR="/usr/local/opt/openssl@1.1"' >> ~/.zshrc
> % echo 'export Qt5_DIR="/usr/local/opt/qt"' >> ~/.zshrc
> % echo 'export Qt5LinguistTools_DIR="/usr/local/opt/qt/lib/cmake/Qt5LinguistTools/"' >> ~/.zshrc
> ```
> 
> > If the reader uses `bash` instead of `zsh`, they would use a different profile than `~/.zshrc`. As an aside, I'm fairly certain the `%` instead of a `$` signifies that the shell is `zsh` rather than `bash`.

I'm fairly certain the % and $ characters have nothing to do with the shell being used. Depends how the prompt is done so we can't be sure of anything.

Is it safe to assume zsh use is widespread on Mac? I'm wondering if those should preferably be written from the perspective of a zsh or a bash user...
 
> > Note: I feel like it would be tidier and more resilient to update CMakeLists.txt so that it uses Homebrew to find these.

Well, as previously discussed, this is not going to happen. ;-)

>     1. Install BasicTex (or MacTex)
>        https://www.tug.org/mactex/morepackages.html
>        Or, direct link:
>        http://mirror.ctan.org/systems/mac/mactex/BasicTeX.pkg

Probably mention here that it is optional. It's really mostly for people wanting to work with the documentation. Maybe not the best thing to push people to install just to roll a patch.

>     2. Install the Packages package creation tool:
>        http://s.sudre.free.fr/Software/Packages/about.html
> 
>     3. Clone the Nextcloud repository to a convenient location, such as `~/Repositories`:
> 
> 
> ```
> % cd ~/Repositories
> ```
> 
> Then, assuming you've already forked Nextcloud Desktop:
> 
> ```
> git clone git@github.com:<github_username>/desktop.git
> ```
> 
> Or, for HTTPS (soon to be deprecated):
> 
> ```
> git clone https://github.com/<github_username>/desktop.git
> ```
> 
> Or, if you haven't forked Nextcloud Desktop, clone as read-only:
> 
> ```
> git clone git@github.com:nextcloud/desktop.git
> ```
> 
> Or, for HTTPS (soon to be deprecated):
> 
> ```
> git clone https://github.com/nextcloud/desktop.git
> ```

I'm not sure I'd go in so much details for the git clone nowadays. But maybe I'm wrong and this is still necessary... unsure.

>     1. Enable git submodules:
> 
> 
> ```
> % cd ~/Repositories/desktop
> % git submodule init
> % git submodule update
> ```
> 
> > I honestly don't know what this does.

We have a couple of dependencies in the third party directory which are available as git submodule (not great for shipping dependencies but that does the trick in that case). Note that this step can likely be simplified with:
```
git submodule update --init
```

Git is clever enough nowadays to know if it needs to do the init step or not.

>     1. Create build directory:
> 
> 
> ```
> % cd ~/Repositories/desktop
> % mkdir build
> ```
> 
>     1. Generate the build files:
> 
> 
> ```
> % cd ~/Repositories/desktop/build
> % cmake .. \
> -DCMAKE_INSTALL_PREFIX=~/nextcloud-desktop-client \
> -DCMAKE_BUILD_TYPE=Debug
> ```

You can likely drop the CMAKE_BUILD_TYPE, Debug is supposed to be the default anyway.

> > IMHO the install location `~/nextcloud-desktop-client` may be less than ideal. I've used `~/Builds` in the past, just like I use `~/Repositories`. We Mac users really like proper title case and everything in our home folders.

Honestly if you feel like changing it to something else: be my guest. I don't mind much. I agree the proposed path right now isn't great, at the same time it's the kind of things I don't expect people to follow blindly. Everyone seems to have their opinion on those matters. :-)

>     1. Compile and install:
> 
> 
> ```
> % make install
> ```
> 
> > Note: I repeatedly used the absolute path after `cd` because it's somewhat likely a reader might go back and retry the steps out of order.

Yes, that's a good idea.

As for the last part "Things to be fixed elsewhere". @FlexW could you check that you integrated (1) and (2) in your build system PR? As for (3) we discussed it before and I don't think this should go in.

**Comment by elsiehupp** [comment]
> I'm fairly certain the % and $ characters have nothing to do with the shell being used. Depends how the prompt is done so we can't be sure of anything.

I googled it, and apparently my supposition was correct (according to [this article](https://eshop.macsales.com/blog/56921-moving-from-bash-to-zsh-terminal-changes-in-macos-catalina/)).

> Is it safe to assume zsh use is widespread on Mac? I'm wondering if those should preferably be written from the perspective of a zsh or a bash user...

Apparently Apple switched the default shell with macOS Catalina [back in 2019](https://www.theverge.com/2019/6/4/18651872/apple-macos-catalina-zsh-bash-shell-replacement-features). (I thought it was earlier, but that may have just been when Apple originally added `zsh`.) People who have upgraded from older versions of macOS may still default to `bash`, though Apple recommends switching over and provides [instructions for doing so](https://support.apple.com/en-us/HT208050), which may be worth linking to when mentioning the assumption that Mac users are on `zsh`.

> Probably mention here that [PdfLatex] is optional. It's really mostly for people wanting to work with the documentation. Maybe not the best thing to push people to install just to roll a patch.

It's my understanding that anything marked "Warning" is optional. I could include brief notes regarding optional dependencies, with the recommendation that people go ahead and install all of them.

> I'm not sure I'd go in so much details for the git clone nowadays. But maybe I'm wrong and this is still necessary... unsure.

As someone who has muddled their way through limited documentation and still has to google things, I think it's nice to have things spelled out to a certain extent, as long as it doesn't get too wordy. The main issue with the Git clone is that the URL does vary, and it could potentially be more confusing to put a single URL that might not work in a given circumstance. I just checked, and apparently what is deprecated is [using a GitHub password for HTTPS Git authentication, in favor of using revocable tokens](https://docs.github.com/en/github/using-git/why-is-git-always-asking-for-my-password). I will try and get the explanation as condensed as possible.

> You can likely drop the CMAKE_BUILD_TYPE, Debug is supposed to be the default anyway.

👍

> Honestly if you feel like changing it to something else: be my guest. I don't mind much. I agree the proposed path right now isn't great, at the same time it's the kind of things I don't expect people to follow blindly. Everyone seems to have their opinion on those matters. :-)

The one thing presumably worth emphasizing is that the cloned repository shouldn't go in a cloud-synced folder (especially iCloud Drive). Other than that, yeah, it doesn't really matter (as long as the permissions are correct).

> Well, as previously discussed, [changing the CMake] is not going to happen. ;-)

Fair enough!

**Comment by elsiehupp** [comment]
One question: now that I have write access for the Wiki, should I go ahead and post the Mac instructions? I can do rough drafts in Issues for any other substantial Wiki rewrites I might do, since GitHub doesn't allow forking for the Wikis.

**Comment by er-vin** [comment]
> > I'm fairly certain the % and $ characters have nothing to do with the shell being used. Depends how the prompt is done so we can't be sure of anything.
> 
> I googled it, and apparently my supposition was correct (according to [this article](https://eshop.macsales.com/blog/56921-moving-from-bash-to-zsh-terminal-changes-in-macos-catalina/)).

Yes, sorry for being unclear. What I meant is: that might be what you get as default but that's easy to change. Lots of developers do change the prompt.
 
> > Is it safe to assume zsh use is widespread on Mac? I'm wondering if those should preferably be written from the perspective of a zsh or a bash user...
> 
> Apparently Apple switched the default shell with macOS Catalina [back in 2019](https://www.theverge.com/2019/6/4/18651872/apple-macos-catalina-zsh-bash-shell-replacement-features). (I thought it was earlier, but that may have just been when Apple originally added `zsh`.) People who have upgraded from older versions of macOS may still default to `bash`, though Apple recommends switching over and provides [instructions for doing so](https://support.apple.com/en-us/HT208050), which may be worth linking to when mentioning the assumption that Mac users are on `zsh`.

OK, excellent. So let's assume zsh, that seems like the right choice.

> > Probably mention here that [PdfLatex] is optional. It's really mostly for people wanting to work with the documentation. Maybe not the best thing to push people to install just to roll a patch.
> 
> It's my understanding that anything marked "Warning" is optional. I could include brief notes regarding optional dependencies, with the recommendation that people go ahead and install all of them.

Yes, with @FlexW changes I expect those warnings to be less prominent and to have a nice summary about what is optional or not out of cmake itself. Should make things less confusing hopefully.

> > I'm not sure I'd go in so much details for the git clone nowadays. But maybe I'm wrong and this is still necessary... unsure.
> 
> As someone who has muddled their way through limited documentation and still has to google things, I think it's nice to have things spelled out to a certain extent, as long as it doesn't get too wordy. The main issue with the Git clone is that the URL does vary, and it could potentially be more confusing to put a single URL that might not work in a given circumstance. I just checked, and apparently what is deprecated is [using a GitHub password for HTTPS Git authentication, in favor of using revocable tokens](https://docs.github.com/en/github/using-git/why-is-git-always-asking-for-my-password). I will try and get the explanation as condensed as possible.

Fair enough. Let's keep it like this then.

> > You can likely drop the CMAKE_BUILD_TYPE, Debug is supposed to be the default anyway.
> 
> +1
> 
> > Honestly if you feel like changing it to something else: be my guest. I don't mind much. I agree the proposed path right now isn't great, at the same time it's the kind of things I don't expect people to follow blindly. Everyone seems to have their opinion on those matters. :-)
> 
> The one thing presumably worth emphasizing is that the cloned repository shouldn't go in a cloud-synced folder (especially iCloud Drive). Other than that, yeah, it doesn't really matter (as long as the permissions are correct).

Agreed.

> > Well, as previously discussed, [changing the CMake] is not going to happen. ;-)
> 
> Fair enough!

**Comment by er-vin** [comment]
> > One question: now that I have write access for the Wiki, should I go ahead and post the Mac instructions? I can do rough drafts in Issues for any other substantial Wiki rewrites I might do, since GitHub doesn't allow forking for the Wikis.
> 
> What about updating the official documentation? We have this page in the docs https://docs.nextcloud.com/desktop/3.1/building.html#macos and the content is at https://github.com/nextcloud/desktop/blob/master/doc/building.rst. You will need to create PR's for this and whatever you write there can be copied to the wiki later (eventually perhaps we should have everything in the docs, remove the wiki?)

Oh good point @camilasan thanks for pointing this out... Does it even make sense to have that at both places then? Maybe we should drop the wiki for such build information and use *only* the official documentation then.

**Comment by camilasan** [comment]
> Maybe we should drop the wiki for such build information and use _only_ the official documentation then.

Agreed. We should only have a bit of info in the README and then links to the official documentation because it is usually the first place people look when they get started.
